### ISSUE TITLE
declare strict constructor arguments  - instead of trying to deepseq later

### ISSUE THREAD

**Issue by jwaldmann** [tidalcycles/Tidal#667]
(related: #606 https://github.com/rd--/hosc/pull/2#issuecomment-639522092)

If we do know that we never want lazy values - then we should declare strict constructor arguments. This is much better than sprinkling the code with `rdeepseq` later.  

This states the intent, clearly, and in the right place (when we define the type), and it will check it automatically, every time,  as early as possible (when the constructor is called).

By default, this should apply to
* all data types that have `NFData` instances (currently, `ArcF`, `Context`, `Value`, `Pattern`)
* _all_ data types (e.g., `Tempo`, `State`, `Config`, `EventF`)?
The latter could be achieved by pragma `StrictData` https://gitlab.haskell.org/ghc/ghc/-/wikis/strict-pragma I will give this a try.

What would be the remaining non-strict data? I think
* Pattern (because it contains a function)
* all prelude data types (List, Maybe)

Is there any part of the public API (Tidal's pattern combinators) that relies on laziness?

**Comment by yaxu** [comment]
> Is there any part of the public API (Tidal's pattern combinators) that relies on laziness?

No, is this short answer..

**Comment by jwaldmann** [comment]
travis-CI tells me that StrictData is not available in ghc before 8.

**Comment by yaxu** [comment]
Yes I think it would be best to keept support of older versions of ghc, if it's just a matter of adding some `!`s.

**Comment by ndr-brt** [comment]
If it's important to keep support for ghc 7.x we can add a build on the ci

**Comment by yaxu** [comment]
Centos 6 that is mentioned in that thread is no longer maintained. Ubuntu Xenial is still supported and is on ghc 7.10.3. It might be worth adding that.
